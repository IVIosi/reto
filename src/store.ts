import {Context, default as React} from 'react'
import {Container} from './container'

export interface Store<P extends unknown[] = any[], V = any> {
  (...args: P): V
  defaultValue?: V
  optional?: boolean
  displayName?: string
  // defaultProps?
  
  Context?: Context<Container<V>>
}

export type StoreP<S extends Store> = S extends Store<infer P, any> ? P : never
export type StoreV<S extends Store> = S extends Store<any, infer V> ? V : never

export function getStoreContext<S extends Store>(S: S) {
  if (!S.Context) {
    S.Context = React.createContext(new Container<StoreV<S>>(S.defaultValue))
  }
  return S.Context
}
