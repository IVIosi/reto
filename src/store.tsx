import * as React from 'react'
import {contextSymbol} from './metadata-symbols'
import produce, {createDraft, Draft, finishDraft} from 'immer'

export type ConstructorType<T> = { new (...args: any[]): T }
export type StoreType<T extends Store = Store> = ConstructorType<T>

export class Store<T extends {} = {}> {
  state: T
  // getters: {}
  storeWillDestroy?(): void
  mutate(f: (draft: Draft<T>) => void) {
    this.state = produce(this.state, f)
    this.propagate()
  }
  propagate() {
    for (let subscriber of this.subscribers) {
      subscriber()
    }
  }
  subscribers: Function[] = []
  effects = [
    {
      handler: function () {
      
      },
      deps: function () {
      
      }
    }, {
    },
  ]
}


export function store<T extends {new(...args:any[]):{}}>(constructor: T) {
  Reflect.defineMetadata(contextSymbol, React.createContext(null), constructor)
}


// function useFunctionStore(Store: FunctionStore) {
//   const state = Store.init()
// }
