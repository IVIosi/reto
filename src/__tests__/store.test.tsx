import {Provider, useStore} from '..'
import {BarStore, BarWithoutDecoratorStore, FooStore} from './stores/foo.store'
import * as React from 'react'
import {FC} from 'react'
import * as testing from 'react-testing-library'


test('provider initialize', function () {
  const App: FC = (props) => {
    const fooStore = useStore(FooStore)
    
    function changeStore() {
      fooStore.mutate(draft => {
        draft.x++
      })
    }
    return (
      <div>
        <button onClick={changeStore}>Change</button>
        {fooStore.state.x}
      </div>
    )
  }
  const renderer = testing.render(
    <Provider of={FooStore}>
      <App/>
    </Provider>
  )
  expect(renderer.asFragment()).toMatchSnapshot()
  testing.fireEvent.click(testing.getByText(renderer.container, 'Change'))
  expect(renderer.asFragment()).toMatchSnapshot()
})


test('rerender on dependency update', function () {
  const App: FC = (props) => {
    const barStore = useStore(BarStore)
    
    function changeStore() {
      barStore.fooStore.mutate(draft => {
        draft.x++
      })
    }
    return (
      <div>
        <button onClick={changeStore}>Change</button>
        {barStore.fooStore.state.x}
      </div>
    )
  }
  const renderer = testing.render(
    <Provider of={FooStore}>
      <Provider of={BarStore}>
        <App/>
      </Provider>
    </Provider>
  )
  expect(renderer.asFragment()).toMatchSnapshot()
  testing.fireEvent.click(testing.getByText(renderer.container, 'Change'))
  expect(renderer.asFragment()).toMatchSnapshot()
})


test('without decorator', function () {
  const App: FC = () => {
    const barStore = useStore(BarWithoutDecoratorStore)
    return (
      <div>
        {barStore.fooStore.state.x}
      </div>
    )
  }
  
  const renderer = testing.render(
    <Provider of={FooStore}>
      <Provider of={BarWithoutDecoratorStore}>
        <App/>
      </Provider>
    </Provider>
  )
  
  expect(renderer.asFragment()).toMatchSnapshot()
})
