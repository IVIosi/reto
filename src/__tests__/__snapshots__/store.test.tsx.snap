// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no extra render on children 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change Parent
    </button>
    parent state: 1
    <div>
      <button>
        Change Store
      </button>
      1
    </div>
  </div>
</DocumentFragment>
`;

exports[`no extra render on children 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change Parent
    </button>
    parent state: 1
    <div>
      <button>
        Change Store
      </button>
      2
    </div>
  </div>
</DocumentFragment>
`;

exports[`no extra render on children 3`] = `
<DocumentFragment>
  <div>
    <button>
      Change Parent
    </button>
    parent state: 2
    <div>
      <button>
        Change Store
      </button>
      2
    </div>
  </div>
</DocumentFragment>
`;

exports[`no extra render on children 4`] = `
<DocumentFragment>
  <div>
    <button>
      Change Parent
    </button>
    parent state: 3
    <div>
      <button>
        Change Store
      </button>
      2
    </div>
  </div>
</DocumentFragment>
`;

exports[`provider initialize 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;

exports[`provider initialize 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    2
  </div>
</DocumentFragment>
`;

exports[`provider initialize with args 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    5
  </div>
</DocumentFragment>
`;

exports[`provider initialize with args 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    6
  </div>
</DocumentFragment>
`;

exports[`rerender on dependency update 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;

exports[`rerender on dependency update 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    3
  </div>
</DocumentFragment>
`;

exports[`useStoreState 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;

exports[`useStoreState 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    2
  </div>
</DocumentFragment>
`;
